### PR TITLE
Migrated to latest Microsoft Graph API

### PR DESCRIPTION
**Pull request by jebeaudet**
Everything is updated to the latest Microsoft Graph API.

I removed a bunch of dead code that wasn't used and added caching for the user profile
J:FOUND-4885

### CODE REVIEW

File: src/main/java/org/springframework/social/office365/api/domain/User.java [hunk 1]
```diff
@@ -12,107 +15,64 @@
...
-    public String getId()
+    private final String id;
+    private final String givenName;
+    private final String surname;
+    private final String displayName;
+    private final String userPrincipalName;
+    private final String email;
+    private final Set<String> aliases;
+
+    @JsonCreator
```
Comment (jebeaudet) [inline src/main/java/org/springframework/social/office365/api/domain/User.java:26]: Immutable object for caching

File: src/main/java/org/springframework/social/office365/api/impl/UserTemplate.java [hunk 2]
```diff
@@ -26,17 +27,14 @@ public UserTemplate(RestTemplate restTemplate, boolean isAuthorized, String base
     @Override
     public User getUserProfile()
     {
-        checkAuthorization(isAuthorized);
-
-        return restTemplate.getForObject(URI.create(url + "/me?api-version=2013-11-08"), User.class);
-    }
+        if (user != null) {
```
Comment (jebeaudet) [inline src/main/java/org/springframework/social/office365/api/impl/UserTemplate.java:30]: user profile dummy cache

File: src/main/java/org/springframework/social/office365/connect/Office365OAuth2Template.java [hunk 3]
```diff
@@ -21,11 +21,11 @@
...
-    public static final String AUTHORIZE_URI = "https://login.microsoftonline.com/common/oauth2/authorize";
-    public static final String ACCESS_TOKEN_URI = "https://login.microsoftonline.com/common/oauth2/token";
-
-    public static final String RESOURCE_TO_ACCESS_URI = "https://graph.windows.net";
-    public static final String RESOURCE_KEY = "resource";
+    public static final String AUTHORIZE_URI = "https://login.microsoftonline.com/common/oauth2/v2.0/authorize";
+    public static final String ACCESS_TOKEN_URI = "https://login.microsoftonline.com/common/oauth2/v2.0/token";
+    private static final String SCOPE_PARAMETER = "scope";
+    // https://learn.microsoft.com/en-us/azure/active-directory/develop/scopes-oidc#the-default-scope
+    private static final String ALL_SCOPES_REQUESTED_IN_AUTHORIZATION_STEP = "https://graph.microsoft.com/.default";
```
Comment (jebeaudet) [inline src/main/java/org/springframework/social/office365/connect/Office365OAuth2Template.java:28]: magic scope to get all the scope requested during the authn request

File: src/main/java/org/springframework/social/office365/connect/Office365OAuth2Template.java [hunk 4]
```diff
@@ -38,16 +38,6 @@ public Office365OAuth2Template(String clientId,
         setUseParametersForClientAuthentication(true);
     }
 
-    public Office365OAuth2Template(String clientId,
```
Comment (jebeaudet) [inline src/main/java/org/springframework/social/office365/connect/Office365OAuth2Template.java:41]: dead code

File: src/main/java/org/springframework/social/office365/connect/Office365OAuth2Template.java [hunk 5]
```diff
@@ -56,10 +46,7 @@ public AccessGrant exchangeForAccess(String authorizationCode,
         if (additionalParameters == null) {
             additionalParameters = new LinkedMultiValueMap<>();
         }
-
-        if (!additionalParameters.containsKey(RESOURCE_KEY)) {
-            additionalParameters.add(RESOURCE_KEY, RESOURCE_TO_ACCESS_URI);
```
Comment (jebeaudet) [inline src/main/java/org/springframework/social/office365/connect/Office365OAuth2Template.java:61]: this is the new way of doing things for the v2.0 endpoint

File: src/main/java/org/springframework/social/office365/connect/Office365OAuth2Template.java [hunk 6]
```diff
@@ -21,11 +21,11 @@
  */
 public class Office365OAuth2Template extends OAuth2Template
 {
-    public static final String AUTHORIZE_URI = "https://login.microsoftonline.com/common/oauth2/authorize";
-    public static final String ACCESS_TOKEN_URI = "https://login.microsoftonline.com/common/oauth2/token";
-
-    public static final String RESOURCE_TO_ACCESS_URI = "https://graph.windows.net";
-    public static final String RESOURCE_KEY = "resource";
+    public static final String AUTHORIZE_URI = "https://login.microsoftonline.com/common/oauth2/v2.0/authorize";
```
Comment (jebeaudet) [inline src/main/java/org/springframework/social/office365/connect/Office365OAuth2Template.java:24]: new authn/token endpoints

File: src/main/java/org/springframework/social/office365/api/impl/Office365Template.java [hunk 7]
```diff
@@ -23,19 +21,14 @@
  */
 public class Office365Template extends AbstractOAuth2ApiBinding implements Office365
 {
-    public static final String SHARE_POINT_BASE_URL = "https://graph.windows.net";
-    public static final String GRAPH_BASE_URL = "https://graph.windows.net";
+    public static final String GRAPH_BASE_URL = "https://graph.microsoft.com/v1.0";
```
Comment (jebeaudet) [inline src/main/java/org/springframework/social/office365/api/impl/Office365Template.java:24]: new API endpoint

File: src/main/java/org/springframework/social/office365/connect/Office365OAuth2Template.java (hunk 3 above)
Comment (fredboutin) [reply src/main/java/org/springframework/social/office365/connect/Office365OAuth2Template.java:28]: wanna see some magic

**Review by fredboutin — APPROVED** [review]

**Pull request merged**